src/txn/txn_timestamp.c

```diff
@@ -740,74 +752,110 @@ __wt_txn_set_timestamp(WT_SESSION_IMPL *session, const char *cfg[])
  */
 int
 __wt_txn_parse_prepare_timestamp(
-    WT_SESSION_IMPL *session, const char *cfg[], wt_timestamp_t *timestamp)
+    WT_SESSION_IMPL *session, const char *cfg[])
 {
 	WT_CONFIG_ITEM cval;
-	WT_TXN *prev;
+	WT_TXN *prev, *txn;
 	WT_TXN_GLOBAL *txn_global;
-	wt_timestamp_t oldest_ts;
-	char ts_string[WT_TS_INT_STRING_SIZE];
+	wt_timestamp_t oldest_ts, timestamp;
+	char ts_string[2][WT_TS_INT_STRING_SIZE];
+	bool checkpoint_accounted, round_to_oldest;
 
+	txn = &session->txn;
 	txn_global = &S2C(session)->txn_global;
 
+	checkpoint_accounted = round_to_oldest = false;
+
 	WT_RET(__wt_config_gets_def(session,
 	    cfg, "prepare_timestamp", 0, &cval));
-	if (cval.len > 0) {
-		if (F_ISSET(&session->txn, WT_TXN_HAS_TS_COMMIT))
-			WT_RET_MSG(session, EINVAL,
-			    "commit timestamp should not have been set before "
-			    "prepare transaction");
+	if (cval.len == 0)
+		WT_RET_MSG(session, EINVAL, "prepare timestamp is required");
 
-		WT_RET(__wt_txn_parse_timestamp(
-		    session, "prepare", timestamp, &cval));
+	if (F_ISSET(&session->txn, WT_TXN_HAS_TS_COMMIT))
+		WT_RET_MSG(session, EINVAL, "commit timestamp "
+		    "should not have been set before prepare transaction");
 
-		/*
-		 * Prepare timestamp must be later/greater than latest active
-		 * read timestamp.
-		 */
-		__wt_readlock(session, &txn_global->read_timestamp_rwlock);
-		prev = TAILQ_LAST(&txn_global->read_timestamph,
-		    __wt_txn_rts_qh);
-		while (prev != NULL) {
-			/*
-			 * Skip any transactions that are not active.
-			 */
-			if (prev->clear_read_q) {
-				prev = TAILQ_PREV(
-				    prev, __wt_txn_rts_qh, read_timestampq);
-				continue;
-			}
-			if (prev->read_timestamp >= *timestamp) {
-				__wt_readunlock(session,
-				    &txn_global->read_timestamp_rwlock);
+	WT_RET(__wt_txn_parse_timestamp(session, "prepare", &timestamp, &cval));
+
+	WT_RET(__wt_config_gets_def(session, cfg, "round_to_oldest", 0, &cval));
+	round_to_oldest = cval.val;
+
+	__wt_readlock(session, &txn_global->read_timestamp_rwlock);
+
+	oldest_ts = txn_global->oldest_timestamp;
+
+	/*
+	 * Prepare timestamp must be later/greater than latest active
+	 * read timestamp, if any.
+	 */
+	prev = TAILQ_LAST(
+	    &txn_global->read_timestamph, __wt_txn_rts_qh);
+	while (prev != NULL) {
+		/* Skip other transactions that are not active. */
+		if (prev->clear_read_q) {
+			prev = TAILQ_PREV(
+			    prev, __wt_txn_rts_qh, read_timestampq);
+			continue;
+		}
+
+		/* Skip self transaction. */
+		if (prev->read_timestamp == txn->read_timestamp)
+			continue;
```

Comment:
Just curious, doesn't this form an infinite loop, since we just continue without moving the pointer to the next element in the queue. Or Am I missing something? . 

Even for line 808(Skip checkpoint timestamp).